src/sandstorm/supervisor.c++

```diff
@@ -1586,14 +1594,195 @@ public:
     //   restart independently of the app, but the `OngoingNotification` provided by the app is
     //   not required to be persistent. The supervisor thus takes care of the complication of
     //   dealing with persistence through front-end restarts.
-//  }
+    sandstorm::isBackgrounded = true;
+    auto params = context.getParams();
+
+    kj::String lockId(kj::heapString(16));
+    randombytes_buf(lockId.begin(), lockId.size());
+    LockInfo lock(kj::mv(lockId));
+    auto insertResult = wakelockMap.insert(std::make_pair(static_cast<kj::StringPtr>(lock.id), kj::mv(lock)));
+    auto lockRef = insertResult.first;
+
+    auto req = sandstormCore.getOwnerNotificationTargetRequest().send().getOwner().addOngoingRequest();
+    req.setDisplayInfo(params.getDisplayInfo());
+    req.setNotification(kj::heap<WrappedOngoingNotification>(params.getNotification(), lockRef->first,  *this));
+    context.releaseParams();
+    return req.send().then([this, lockRef, params, context](auto args) mutable {
+      auto req = args.getHandle().template castAs<SystemPersistent>().saveRequest();
+      auto grainOwner = req.getSealFor().initGrain();
+      grainOwner.setGrainId(grainId);
+      grainOwner.getSaveLabel().setDefaultText("ongoing notification handle");
+      return req.send().then([this, context, lockRef](auto args) mutable {
+        lockRef->second.sturdyRef = kj::heapArray(args.getSturdyRef());
+        context.getResults().setHandle(kj::heap<WakelockHandle>(lockRef->first, *this));
+      });
+    });
+  }
+
+  OngoingNotificationPersistent::Client restoreWakeLockNotification(kj::StringPtr wakelockId) {
+    auto iter = wakelockMap.find(wakelockId);
+    KJ_REQUIRE(iter != wakelockMap.end(), "Wakelock id not found");
+    return kj::heap<WrappedOngoingNotification>(iter->second.ongoingNotification, iter->first, *this);
+  }
+
+  void incrementNotificationRefCount(kj::StringPtr wakelockId) {
+    auto iter = wakelockMap.find(wakelockId);
+    KJ_REQUIRE(iter != wakelockMap.end(), "Tried to increment ref count for a wakelock that doesn't exist");
+    iter->second.notificationRefCount++;
+  }
+
+  void decrementNotificationRefCount(kj::StringPtr wakelockId) {
+    auto iter = wakelockMap.find(wakelockId);
+    if (iter == wakelockMap.end()) {
+      KJ_LOG(WARNING, "Tried to decrement ref count for a wakelock that doesn't exist");
+      return;
+    }
+    iter->second.notificationRefCount--;
+    if (iter->second.notificationRefCount == 0) {
+      tasks.add(iter->second.ongoingNotification.cancelRequest().send().then([](auto args) {}));
+      stopWakelock(iter);
+    }
+  }
+private:
+  class LockInfo {
+  public:
+    kj::String id;
+    kj::Array<byte> sturdyRef;
+    OngoingNotification::Client ongoingNotification;
+    bool isHandleActive;
+    uint32_t notificationRefCount;
+
+    LockInfo(kj::String&& id) :
+      id(kj::mv(id)), ongoingNotification(nullptr), isHandleActive(true), notificationRefCount(0) {
```

Comment:
I find it's a little easier to tell the initializers apart from the function body by writing like:

    LockInfo(kj::String&& id)
        : id(kj::mv(id)), ... {
      SANDSTORM_LOG(...);
      ...
    }